h2/src/main/org/h2/mvstore/MVStore.java

```diff
@@ -776,182 +795,127 @@ private void readStoreHeader() {
-            }
-
-            if (assumeCleanShutdown) {
-                setLastChunk(newest);
-                Queue<Chunk> chunksToVerify = new PriorityQueue<>(20, Collections.reverseOrder(chunkComparator));
-                try {
-                    // load the chunk metadata: although meta's root page resides in the lastChunk,
-                    // traversing meta map might recursively load another chunk(s)
-                    Cursor<String, String> cursor = meta.cursor(DataUtils.META_CHUNK);
-                    while (cursor.hasNext() && cursor.next().startsWith(DataUtils.META_CHUNK)) {
-                        Chunk c = Chunk.fromString(cursor.getValue());
-                        assert c.version <= currentVersion;
-                        // might be there already, due to meta traversal
-                        // see readPage() ... getChunkIfFound()
-                        chunks.putIfAbsent(c.id, c);
-                        chunksToVerify.offer(c);
-                        if (chunksToVerify.size() == 20) {
-                            chunksToVerify.poll();
-                        }
-                    }
-                    Chunk c;
-                    while (assumeCleanShutdown && (c = chunksToVerify.poll()) != null) {
-                        assumeCleanShutdown = readChunkHeaderAndFooter(c.block, c.id) != null;
-                    }
-                } catch(IllegalStateException ignored) {
-                    assumeCleanShutdown = false;
-                }
+        // Step-2: read one chunk of the newest chunk candidates at the end of file, and ignore
+        // the partial write part
+        // bugfix - data lost issue when partial write occurs at end of file store.
+        // @since 2019-07-31 little-pan
+        final Chunk tail = readChunkHeaderAndFooterFromStoreTail();
+        long vdiff;
+        if(tail != null && (vdiff = tail.version - headerVersion) > 0L){
+            if(vdiff < lastCandidates.length){
+                lastCandidates[(int)vdiff] = newest;
+            }
+            newest = tail;
+        }
+        if(newest == null && recoveryMode && tail != null) {
+            headerVersion = tail.version;
+            storeHeader.put(HDR_BLOCK, tail.block);
+            storeHeader.put(HDR_CHUNK, tail.id);
+            storeHeader.put(HDR_VERSION, headerVersion);
+            newest = tail;
+        }
+        if(newest == null){
+            // + Check integrity of the newest chunk and it's version in file header
+            if(headerVersion > 0L && !recoveryMode){
+                throw DataUtils.newIllegalStateException(
+                        DataUtils.ERROR_FILE_CORRUPT, 
+                        "No valid chunk for last version {0}", headerVersion);
             }
+            // Only file headers for new database or in recoveryMode
+            storeHeader.remove(HDR_BLOCK);
+            storeHeader.remove(HDR_CHUNK);
+            storeHeader.remove(HDR_VERSION);
+            return;
```

Comment:
Just return without call to setLastChunk()? What does it mean exactly?

Reply:
It's added.